src/Package/Impl/DataInspect/VariableView.xaml.cs

```diff
@@ -57,16 +71,76 @@ public partial class VariableView : UserControl, IDisposable {
             e.Handled = true;
         }
 
-        private void OnGlobalEnvironmentEvaluation(DebugEvaluationResult result) {
-            var wrapper = new EvaluationWrapper(result);
+        private void EnvironmentProvider_EnvironmentChanged(object sender, REnvironmentChangedEventArgs e) {
+            int selectedIndex = 0;
 
-            var rootNodeModel = new VariableNode(_settings, wrapper);
+            var currentItem = EnvironmentComboBox.SelectedItem as REnvironment;
+            if (currentItem != null && !e.Environments[0].FrameIndex.HasValue) {
+                int index = e.Environments.FindIndex((env) => env.Name == currentItem.Name);
+                selectedIndex = index == -1 ? selectedIndex : index;
+            }
+
+            EnvironmentComboBox.ItemsSource = e.Environments;
+            EnvironmentComboBox.SelectedIndex = selectedIndex;
+        }
 
-            VsAppShell.Current.DispatchOnUIThread(
-                () => {
-                    EnvironmentName.Text = GlobalEnvironmentName;
-                    _rootNode.Model = rootNodeModel;
-                });
+        private void EnvironmentComboBox_SelectionChanged(object sender, SelectionChangedEventArgs e) {
+            if (!object.Equals(EnvironmentComboBox.ItemsSource, _defaultEnvironments) && e.AddedItems.Count > 0) {
+                var env = e.AddedItems[0] as REnvironment;
+                if (env != null) {
+                    SetRootModelAsync(env).SilenceException<Exception>().DoNotWait();
+                }
+            }
+        }
+
+        private async Task<DebugSession> GetDebugSessionAsync() {
+            if (_debugSession != null) {
+                return _debugSession;
+            }
+
+            var debugSessionProvider = VsAppShell.Current.ExportProvider.GetExportedValue<IDebugSessionProvider>();
+            if (_debugSession == null) {
+                Debug.Assert(_rSession != null);
+                _debugSession = await debugSessionProvider.GetDebugSessionAsync(_rSession);
+            }
+
+            return _debugSession;
+        }
+
+        private async Task SetRootModelAsync(REnvironment env) {
+            await TaskUtilities.SwitchToBackgroundThread();
+
+            var debugSession = await GetDebugSessionAsync();
+            var frames = await debugSession.GetStackFramesAsync();
+            var frame = frames.FirstOrDefault(f => f.Index == 0);
+
+            if (frame != null) {
+                const DebugEvaluationResultFields fields = DebugEvaluationResultFields.Classes
+                        | DebugEvaluationResultFields.Expression
+                        | DebugEvaluationResultFields.TypeName
+                        | (DebugEvaluationResultFields.Repr | DebugEvaluationResultFields.ReprStr)
+                        | DebugEvaluationResultFields.Dim
+                        | DebugEvaluationResultFields.Length;
+                DebugEvaluationResult result = await frame.EvaluateAsync(
```

Comment:
If the sole purpose of obtaining the stack frames above is to call `EvaluateAsync` here and get `DebugEvaluationResult`, then this isn't actually necessary. `DebugSession` has `EvaluateAsync` that does the same, but lets you specify the environment explicitly (as an expression string, so it can be anything, e.g. `".GlobalEnv"`, and not just a frame).